### PR TITLE
Docs - OIDC instructions

### PR DESCRIPTION
**Pull request by gongjason**
- added Azure aD
- updated Google + Okta

updates docs for changes made in https://github.com/firezone/firezone/pull/586

Todo
- [x] Add upgrade section noting that admins must revoke all previously granted tokens (or delete the OAuth client app and recreate) to be given a new `refresh_token` used for disabling VPN when the user's token can't be refreshed.

### CODE REVIEW

**Comment by princemaple** [comment]
Should we add it now or later:

Google gives you `refresh_token` no matter what
Microsoft and Okta requires `offline_access` scope to get the `refresh_token`

They will be required for the auto disable functionality.

**Comment by gongjason** [comment]
@princemaple it prob makes sense to add it together with the auto-disable functionality (if it isn't merged yet). It's pretty easy to update scopes/permissions for Azure/Okta.

**Comment by princemaple** [comment]
Nah it's not merged. It's going to take while to finish. Cool. I just thought that it'd be a widely wanted feature so users might as well get ready for it already.

File: docs/docs/authenticate/azure-ad.md [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+Firezone supports Single Sign-On (SSO) using Azure Active Directory through the generic
+OIDC connector. This guide will walk you through how to obtain the following
+config settings required for the integration:
+
+1. `discovery_document_uri`: This URL returns a JSON with information to
+construct a request to the OpenID server.
+1. `client_id`: The client ID of the application.
+1. `client_secret`: The client secret of the application.
+1. `redirect_uri`: Instructs OIDC identity provider where the response to the
```
Comment (jamilbk) [inline docs/docs/authenticate/azure-ad.md]: clarif:

```
1. `redirect_uri`: Instructs OIDC provider where to redirect after authentication. Should be your firezone `EXTERNAL_URL + /auth/oidc/<provider_key>/callback`, e.g. `https://firezone.example.com/auth/oidc/azure/callback`
```

File: docs/docs/authenticate/azure-ad.md [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+request should be sent.
+1. `response_type`: Set to `code`.
+1. `scope`: The permissions required by Firezone.
```
Comment (jamilbk) [inline docs/docs/authenticate/azure-ad.md]: suggested technical clarification:

```
1. `scope`: [OIDC scopes](https://openid.net/specs/openid-connect-basic-1_0.html#Scopes) to obtain from your OIDC provider. This should be set to `openid email` to provide Firezone with the user's email in the returned claims.
```

File: docs/docs/authenticate/index.md [hunk 3]
```diff
@@ -5,21 +5,61 @@ nav_order: 3
...
+
+By default, Firezone uses local email/password authentication, but can also
+support integration with any generic OpenID Connect
+(OIDC) identity provider. This allows users to sign in to Firezone using
+their credentials from Okta, Google, Azure AD, or your own custom identity provider.
+
+## Integrating A Generic OIDC Identity Provider
+
+The example below details the config settings required by Firezone to enable SSO
+through an OIDC identity provider. The configuration file can be found at
```
Comment (jamilbk) [inline docs/docs/authenticate/index.md]: nit: `OIDC identity provider` -> `OIDC provider`? I think this is the commonly used terminology around the web. `OIDC consumer` would be Firezone

File: docs/docs/authenticate/index.md [hunk 4]
```diff
@@ -5,21 +5,61 @@ nav_order: 3
...
+`/etc/firezone/firezone.rb`. To pick up changes, run `firezone-ctl reconfigure`
+and `firezone-ctl restart` to update the application.
+
+```ruby
+# This is an example using Google as an SSO identity provider. 
+# Multiple OIDC configs can be added to the same Firezone instance.
+
+default['firezone']['authentication']['oidc'] = {
+  google: [
```
Comment (jamilbk) [inline docs/docs/authenticate/index.md]: `[` should be `{`

File: docs/docs/authenticate/index.md [hunk 5]
```diff
@@ -5,21 +5,61 @@ nav_order: 3
...
+    discovery_document_uri: "https://accounts.google.com/.well-known/openid-configuration",
+    client_id: "CLIENT_ID",
+    client_secret: "CLIENT_SECRET",
+    redirect_uri: "https://firezone.example.com/auth/oidc/google/callback",
+    response_type: "code",
+    scope: "openid email profile",
+    label: "Google"
+  ]
```
Comment (jamilbk) [inline docs/docs/authenticate/index.md]: should be `}`

File: docs/docs/authenticate/index.md [hunk 6]
```diff
@@ -5,21 +5,61 @@ nav_order: 3
...
+}
+```
+
+We've included instructions on how to set up Firezone with several popular
+identity providers:
```
Comment (jamilbk) [inline docs/docs/authenticate/index.md:71]: nit: `OIDC providers` ?

Comment (jamilbk) [reply docs/docs/authenticate/index.md:71]: Ah nvmd, I see now

File: docs/docs/authenticate/google.md [hunk 7]
```diff
@@ -59,39 +75,38 @@ On the OAuth client ID creation screen:
...
-```
-
-Run the following commands to apply the changes:
-
-```text
-firezone-ctl reconfigure
-firezone-ctl restart
+# Using Google as the SSO identity provider
+default['firezone']['authentication']['oidc'] = {
+  google: [
```
Comment (jamilbk) [inline docs/docs/authenticate/google.md]: `[` -> `{`

File: docs/docs/authenticate/google.md [hunk 8]
```diff
@@ -59,39 +75,38 @@ On the OAuth client ID creation screen:
...
+    discovery_document_uri: "https://accounts.google.com/.well-known/openid-configuration",
+    client_id: "CLIENT_ID",
+    client_secret: "CLIENT_SECRET",
+    redirect_uri: "https://firezone.example.com/auth/oidc/google/callback",
+    response_type: "code",
+    scope: "openid email profile",
+    label: "Google"
+  ]
```
Comment (jamilbk) [inline docs/docs/authenticate/google.md]: `]` -> `}`

File: docs/docs/authenticate/google.md [hunk 9]
```diff
@@ -9,16 +9,32 @@ description: >
...
+1. `response_type`: Set to `code`.
+1. `scope`: The permissions required by Firezone.
+This should be set to `openid email profile`.
+1. `label`: The button label text that shows up on your Firezone login screen.
 
 ![Firezone Google SSO Login](https://user-images.githubusercontent.com/52545545/156853456-1ab3f041-1002-4c79-a266-82acb5802890.gif){:width="600"}
 
+_Note: Previously, Firezone used pre-configured Oauth2 providers. We've moved to
+OIDC based authentication, which allows for any OpenID Connect provider
+(Google, Okta, Dex) to be used for authetication._
```
Comment (jamilbk) [inline docs/docs/authenticate/google.md]: typo: `authetication`

File: docs/docs/authenticate/google.md [hunk 10]
```diff
@@ -9,16 +9,32 @@ description: >
...
+
```
Comment (jamilbk) [inline docs/docs/authenticate/google.md:39]: Maybe add a paragraph here strongly encouraging users to update to the new generic OIDC provider format?

> We strongly recommending transitioning your existing Google or Okta-based SSO configuration to the generic OIDC-based configuration format described here. We'll be removing the Google-specific and Okta-specific SSO functionality in a future release.

File: docs/docs/authenticate/azure-ad.md (hunk 2 above)
Comment (jamilbk) [reply docs/docs/authenticate/azure-ad.md]: Nevermind, we can leave `profile` in there as it's fairly common and will be useful for sending outbound emails.

**Review by jamilbk — CHANGES_REQUESTED** [review]
Just some minor nitpicks and typos!

File: docs/docs/authenticate/index.md (hunk 3 above)
Comment (gongjason) [reply docs/docs/authenticate/index.md]: 👍  good to know!

File: docs/docs/administer/upgrade.md [hunk 11]
```diff
@@ -25,6 +25,40 @@ To upgrade Firezone, follow these steps:
...
+[create a new app integration]({%link docs/authenticate/index.md%}).
+
+Upgrading to >= 0.3.16 requires the `offline_access` scope for OIDC integrations.
+This ensures Firezone syncs with the identity provider and VPN access is terminated
+once the user is removed.
+
+The refresh token required to sync users with your identity
+provider can only be obtained the first time a user grants permissions to Firezone.
+So even if the scope was provided previously, you will need to re-setup the OIDC
+integration after upgrading.
```
Comment (jamilbk) [inline docs/docs/administer/upgrade.md]: How does one re-setup the OIDC integration? I think we may get some confused users here. Either the Oauth app needs to be completely removed and recreated (new client_id and client_secret) or the user needs to revoke the permissions for the app (I think this varies on a case-by-case).

Option (1) is the one that should work in most cases.

File: docs/docs/administer/upgrade.md [hunk 12]
```diff
@@ -25,6 +25,46 @@ To upgrade Firezone, follow these steps:
 Occasionally problems arise. If you hit any, please let us know by [filing an
 issue](https://github.com/firezone/firezone/issues/new/choose).
 
+## Upgrading from 0.3.x to >= 0.3.16
+
+Follow the instructions below based on your current version and setup:
+
+### I have an existing OIDC integration
```
Comment (gongjason) [inline docs/docs/administer/upgrade.md:32]: @jamilbk @princemaple what do you think about this?

File: docs/docs/administer/upgrade.md [hunk 13]
```diff
@@ -25,6 +25,46 @@ To upgrade Firezone, follow these steps:
...
+
+If Firezone is able to successfully retrieve the refresh token, you will see
+the **OIDC Connections** heading in the user details page of the web UI for
+users authenticated through your OIDC provider.
+
+![OIDC Connections](https://user-images.githubusercontent.com/52545545/173169922-b0e5f2f1-74d5-4313-b839-6a001041c07e.png)
+
+If this does not work, you will need to delete your existing OAuth app
+and repeat the OIDC setup steps to
+[create a new app integration]({%link docs/authenticate/index.md%}).
```
Comment (gongjason) [inline docs/docs/administer/upgrade.md:54]: @princemaple @jamilbk what do you think about this?

Comment (princemaple) [reply docs/docs/administer/upgrade.md:54]: Maybe worth a mention that they can turn on the site wide expiry (if they were set to Never previously), so they can make sure that the removed users are properly handled.

Comment (jamilbk) [reply docs/docs/administer/upgrade.md:54]: I think this is clear enough. Let's go with this and if we get any confusion around this we can add more clarification here.

**Review by jamilbk — APPROVED** [review]
Looks awesome, very detailed and clear. Much much improved, thanks!

**Pull request merged**